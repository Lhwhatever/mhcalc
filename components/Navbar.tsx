import {
    AppBar,
    Divider,
    IconButton,
    List,
    ListItem,
    ListSubheader,
    SwipeableDrawer,
    Toolbar,
    Typography
} from '@material-ui/core'
import ListItemText from '@material-ui/core/ListItemText'
import MenuIcon from '@material-ui/icons/Menu'
import React from 'react'
import { wrapNextLink } from './AsNextLink'
import endpoints from '../endpoints'

const Navbar = (): JSX.Element => {
    const [drawerOpen, setDrawerOpen] = React.useState(false)

    const toggleDrawer = (open) => (event) => {
        if (event && event.type === 'keydown' && (event.key === 'Tab' || event.key === 'Shift')) return
        setDrawerOpen(open)
    }

    const handleDrawerOpen = toggleDrawer(true)
    const handleDrawerClose = toggleDrawer(false)

    return (
        <React.Fragment>
            <AppBar position="static">
                <Toolbar>
                    <IconButton edge="start" color="inherit" aria-label="menu" onClick={handleDrawerOpen}>
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6">MouseHunt Calculator</Typography>
                </Toolbar>
            </AppBar>
            <SwipeableDrawer anchor="left" open={drawerOpen} onClose={handleDrawerClose} onOpen={handleDrawerOpen}>
                <div>
                    <List component="nav" aria-label="navigation drawer">
                        <Divider />
                        {wrapNextLink(
                            <ListItem button>
                                <ListItemText>Home</ListItemText>
                            </ListItem>,
                            '/'
                        )}
                        <Divider />
                        <List
                            component="div"
                            aria-labelled-by="calculator-list-header"
                            subheader={
                                <ListSubheader component="div" id="calculator-list-header">
                                    Calculators
                                </ListSubheader>
                            }
                        >
                            {endpoints.calculators.map((e) =>
                                wrapNextLink(
                                    <ListItem button key={e.path}>
                                        <ListItemText>{e.name}</ListItemText>
                                    </ListItem>,
                                    e.path
                                )
                            )}
                        </List>
                        <Divider />
                    </List>
                </div>
            </SwipeableDrawer>
        </React.Fragment>
    )
}

export default Navbar
