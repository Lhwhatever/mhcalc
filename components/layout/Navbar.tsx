import { AppBar, AppBarProps, IconButton, Toolbar, Typography } from '@material-ui/core'
import MenuIcon from '@material-ui/icons/Menu'
import React from 'react'

interface BaseNavbarProps {
    showDrawerIcon?: boolean
    onDrawerOpen: () => void
}

export type NavbarProps = BaseNavbarProps & AppBarProps

const Navbar = (props: NavbarProps): JSX.Element => {
    const { showDrawerIcon = true, onDrawerOpen, ...other } = props

    return (
        <AppBar position="fixed" {...other}>
            <Toolbar>
                {showDrawerIcon && (
                    <IconButton edge="start" color="inherit" aria-label="menu" onClick={onDrawerOpen}>
                        <MenuIcon />
                    </IconButton>
                )}
                <Typography variant="h6">MouseHunt Calculator</Typography>
            </Toolbar>
        </AppBar>
    )
}

export default Navbar
