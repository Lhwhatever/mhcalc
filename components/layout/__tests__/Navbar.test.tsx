import { AppBar, IconButton, Toolbar } from '@material-ui/core'
import { createShallow } from '@material-ui/core/test-utils'
import MenuIcon from '@material-ui/icons/Menu'
import React from 'react'
import Navbar from '../Navbar'

describe('Navbar tests', () => {
    let shallow
    let wrapper

    beforeAll(() => {
        shallow = createShallow()
    })

    beforeEach(() => {
        wrapper = shallow(<Navbar onDrawerOpen={jest.fn()} />)
    })

    it('should use AppBar and Toolbar', () => {
        expect(wrapper.find(AppBar)).toHaveLength(1)
        expect(wrapper.find(Toolbar)).toHaveLength(1)
    })

    it('should render menu button only when showDrawerIcon is true or unspecified', () => {
        const expectFindIconButtonMatchesElement = () =>
            expect(
                wrapper.find(IconButton).matchesElement(
                    <IconButton>
                        <MenuIcon />
                    </IconButton>
                )
            )

        expectFindIconButtonMatchesElement().toBe(true)

        wrapper.setProps({ showDrawerIcon: true })
        expectFindIconButtonMatchesElement().toBe(true)

        wrapper.setProps({ showDrawerIcon: false })
        expect(wrapper.find(IconButton)).toHaveLength(0)
    })

    it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot()

        const wrapperNoShowIcon = shallow(<Navbar showDrawerIcon={false} onDrawerOpen={jest.fn()} />)
        expect(wrapperNoShowIcon).toMatchSnapshot()
    })
})
