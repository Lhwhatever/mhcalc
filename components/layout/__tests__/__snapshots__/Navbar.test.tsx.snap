// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar tests should render correctly 1`] = `
<WithStyles(ForwardRef(AppBar))
  position="fixed"
>
  <WithStyles(ForwardRef(Toolbar))>
    <WithStyles(ForwardRef(IconButton))
      aria-label="menu"
      color="inherit"
      edge="start"
      onClick={[MockFunction]}
    >
      <MenuIcon />
    </WithStyles(ForwardRef(IconButton))>
    <WithStyles(ForwardRef(Typography))
      variant="h6"
    >
      MouseHunt Calculator
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;

exports[`Navbar tests should render correctly 2`] = `
<WithStyles(ForwardRef(AppBar))
  position="fixed"
>
  <WithStyles(ForwardRef(Toolbar))>
    <WithStyles(ForwardRef(Typography))
      variant="h6"
    >
      MouseHunt Calculator
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
